feat(observability): API health checks + GraphQL

leebenson opened the pull request:
Initial integration of the API server. Closes #1073. Closes #3225.

### Motivation

This forms the foundation of making a Vector instance observable, and exposing metrics via `vector top` (#3211) and wider work on an upcoming Vector UI.

### Demo

https://www.loom.com/share/a2a83e4a6c0847b7b820af8b439c23f7

### Features

- Groundwork for GraphQL subscriptions, to make internal metrics observable (#3225, #3226.)
- `/playground` GraphQL playground
- `/health` endpoint, returning 200 response + `{"ok": true}`
- `/graphql` GraphQL endpoint, taking the following health query:

```gql
{
  health
}
```

And responding with:

```json
{
  "data": {
    "health": true
  }
}
```

### Caveats / blockers

~Due to the use of `weak_into_raw` in [Juniper](https://github.com/graphql-rust/juniper), the current Rust 1.44 toolchain is incompatible.~ Edit: Switched to async-graphql for stability.

### TODOS

- [x] Move to a cargo workspace crate in `lib/api`. Note: later metrics subscriptions will rely on #3357 
- [x] Add tests. This should include an integration test for spawning the server and asserting health checks.
- [x] Consider whether it's worth returning a `JoinHandle` from `api.stop()`, instead of the current fire-and-forget for graceful shutdown of the API (may be overkill) Update: I don't think it's worth it. Punting.
- [x] Punt on TLS for now; tackle later (possibly moving `tls` to a separate crate?)
- [x] Add a health subscription (in the form of `heartbeat`)

### Later

- Enable TLS binding. Separately? Globally?
- Improve error handling. Currently binding to an existing port will panic.
- Make `/graphql` + `/playground` configurable

leebenson committed:
Update mod.rs

leebenson committed:
Merge branch 'master' into leebenson/api

leebenson committed:
Merge branch 'master' into leebenson/api

leebenson committed:
config diff

jszwedko committed:
docs(http sink): Fix default for rate_limit_num (#3498)

The actual default is 10, not 1000.

Signed-off-by: Jesse Szwedko <jesse@szwedko.me>

fanatid committed:
chore: remove tower:0.1 (#3481)

Signed-off-by: Kirill Fomichev <fanatid@ya.ru>

juchiast committed:
fix(sinks): Wrap StreamSink around TcpSink (#3495)

Signed-off-by: Duy Do <juchiast@gmail.com>

ktff committed:
chore(topology): Wait for detached sink to finish  (#3450)

* Add fix

Signed-off-by: Kruno Tomola Fabro <krunotf@gmail.com>

* Bump

Signed-off-by: ktf <krunotf@gmail.com>

* Propagate panic

Signed-off-by: Kruno Tomola Fabro <krunotf@gmail.com>

* Propagate panic

Signed-off-by: Kruno Tomola Fabro <krunotf@gmail.com>

Hoverbear committed:
chore(docker source): Provision a /var/lib/vector in docker releases (#3492)

Signed-off-by: Ana Hobden <operator@hoverbear.org>

ktff committed:
enhancement: Add events for starting, stopping, and reloading (#3476)

* Add process events

Signed-off-by: ktf <krunotf@gmail.com>

* Use past tense

Signed-off-by: Kruno Tomola Fabro <krunotf@gmail.com>

* One last metric capture

Signed-off-by: Kruno Tomola Fabro <krunotf@gmail.com>

* One more move

Signed-off-by: Kruno Tomola Fabro <krunotf@gmail.com>

* Remove unused use

Signed-off-by: Kruno Tomola Fabro <krunotf@gmail.com>

ktff committed:
enhancement: Add Heartbeat (#3502)

* Beep

Signed-off-by: Kruno Tomola Fabro <krunotf@gmail.com>

* Comment

Signed-off-by: Kruno Tomola Fabro <krunotf@gmail.com>

leebenson committed:
wip

Hoverbear committed:
chore(add_tags transform): Add events to add_tags transform (#3504)

* chore(add_tags transform): Add events to add_tags transform

Signed-off-by: Ana Hobden <operator@hoverbear.org>

* Use refs

Signed-off-by: Ana Hobden <operator@hoverbear.org>

* fmt

Signed-off-by: Ana Hobden <operator@hoverbear.org>

dependabot[bot] committed:
chore(deps): bump pin-project from 0.4.22 to 0.4.23 (#3294)

Bumps [pin-project](https://github.com/taiki-e/pin-project) from 0.4.22 to 0.4.23.
- [Release notes](https://github.com/taiki-e/pin-project/releases)
- [Changelog](https://github.com/taiki-e/pin-project/blob/master/CHANGELOG.md)
- [Commits](https://github.com/taiki-e/pin-project/compare/v0.4.22...v0.4.23)

Signed-off-by: dependabot[bot] <support@github.com>

Co-authored-by: dependabot[bot] <49699333+dependabot[bot]@users.noreply.github.com>

dependabot[bot] committed:
chore(deps): bump indexmap from 1.3.1 to 1.5.1 (#3380)

Bumps [indexmap](https://github.com/bluss/indexmap) from 1.3.1 to 1.5.1.
- [Release notes](https://github.com/bluss/indexmap/releases)
- [Commits](https://github.com/bluss/indexmap/compare/1.3.1...1.5.1)

Signed-off-by: dependabot[bot] <support@github.com>

Co-authored-by: dependabot[bot] <49699333+dependabot[bot]@users.noreply.github.com>

leebenson committed:
graphql init

leebenson committed:
server.stop()

leebenson committed:
api -> lib/api

leebenson committed:
minor cleanup

leebenson committed:
minor cleanup

leebenson committed:
Merge branch 'master' into leebenson/api

leebenson committed:
Merge branch 'master' into leebenson/api

leebenson committed:
reset warp

leebenson committed:
reset warp

MOZGIII committed:
feat: Allow configuring flags via env vars (#2149)

Co-authored-by: Ana Hobden <operator@hoverbear.org>
Co-authored-by: James Turnbull <james@lovedthanlost.net>

leebenson committed:
Merge branch 'master' into leebenson/api

leebenson committed:
Merge branch 'master' into leebenson/api

leebenson committed:
heartbeat subscription

leebenson committed:
heartbeat subscription

leebenson committed:
async-graphql

leebenson committed:
async-graphql

leebenson committed:
appease clippy

leebenson committed:
appease clippy

perfectayush committed:
enhancement(file sink): Add support for gzip compression (#3373)

* enhancement(file sink): Add support for gzip compression

File sink currently doesn't support compression, this rudimentary change will
allow to use compression in file sink

Signed-off-by: Ayush Goyal <ayush@helpshift.com>

* Fix type for compression field in FileSink struct

FileSink struct's compression field was earlier initialized as a bool. Using
Compression enum directly is a better abstraction, simplifies extending FileSink
for other compression types in future

Signed-off-by: Ayush Goyal <ayush@helpshift.com>

* Flush files at shutdown in file sink

File sink should flush files at shutdown not just when files are expired. Not
doing so may result in data loss.

Signed-off-by: Ayush Goyal <ayush@helpshift.com>

* Make linters and formatters happy

* Move `use async_compression` to recommended place by rustfmt
* Fix whitespace

Signed-off-by: Ayush Goyal <ayush@helpshift.com>

* Add tests for gzip file sink

Signed-off-by: Ayush Goyal <ayush@helpshift.com>

binarylogic committed:
docs(file sink): Add `compression` option to the `file` source (#3536)

Signed-off-by: binarylogic <bjohnson@binarylogic.com>

jamtur01 committed:
chore(internal docs): Added some error checking for Make scripts (#3533)

fanatid committed:
chore: less work on shutdown test (#3526)

Signed-off-by: Kirill Fomichev <fanatid@ya.ru>

leebenson committed:
feature="api"

leebenson committed:
feature="api"

leebenson committed:
Merge branch 'master' into leebenson/api

leebenson committed:
Merge branch 'master' into leebenson/api

leebenson committed:
Update Cargo.lock

leebenson committed:
Update Cargo.lock

leebenson committed:
Update config

leebenson committed:
Update config

leebenson committed:
Update Cargo.toml

leebenson committed:
Update Cargo.toml

JeanMertz commented on Cargo.toml:
```diff
@@ -131,14 +135,14 @@ uuid = { version = "0.7", features = ["serde", "v4"], optional = true }
 maxminddb = { version = "0.14.0", optional = true }
 strip-ansi-escapes = { version = "0.1.0"}
 colored = "1.9"
-warp = { version = "0.2.4", default-features = false, optional = true }
```
Perhaps we can only enable the default features we actually need? (unless we need them all?)

leebenson replied:
Yeah, that sounds good. I removed `default-features = false` temporarily to instrument Warp in the API, with the intention of circling back and testing for what we actually need. Thanks for the reminder.

leebenson replied:
FWIW, adding `default-features = false` back to the warp definition made no difference to `cargo build`-- it compiled the cached version just the same. Since we're not also specifying a definitive `features = []` here, I'm not clear on if there's a distinction.

leebenson committed:
clippy

leebenson committed:
clippy

JeanMertz commented on Cargo.toml:
```diff
@@ -231,6 +235,9 @@ wasm-timings = ["wasm"]
 kubernetes = ["k8s-openapi", "evmap"]
 
+# Enables the API
+api = ["warp", "async-graphql", "async-graphql-warp"]
```
Perhaps we can come up with a more descriptive name for this feature? `internal-events-api` or something?

leebenson replied:
There's probably a better name for this than just 'api', although my starting intention is that this would become the de facto gateway into general UI <-> Vector observability and grow into more than just events.

One current use-case we're planning for, for example, is retrieving/validating _Cargo.toml_.

leebenson replied:
To conclude this point: In the absence of a better name, I'll keep this as `api` for now. I do think it'll grow to be more than simply metrics, so trying to keep the nomenclature in sync with its function might be a bit tough; this generalisation may be preferred.

JeanMertz commented on src/api/handler.rs:
```diff
@@ -0,0 +1,13 @@
+
+// health handler, responds with { ok: true }
+pub async fn health() -> Result<impl Reply, Rejection> {
+    Ok(json(&Health { ok: true }))
```
I'm assuming this will eventually expose the actual health of the Vector instance by querying the health of the different active components?

leebenson replied:
Probably, yeah. Right now, this just asserts that the GraphQL handler is alive. There's also a `/health` that returns `{"ok": true}` + a 200 response.

leebenson commented:
Updated to enable the server on `feature = "api"`. Added to defaults.

JeanMertz commented on src/api/schema.rs:
```diff
@@ -0,0 +1,45 @@
+
+#[Subscription]
+impl Subscription {
+    async fn heartbeat(
```
Is there some way to convert Rust doc comments into GraphQL API documentation?

Would be nice to explicitly mention that the interval is defined in milliseconds (and extend that documentation to any future subscriptions/queries/mutations).

leebenson replied:
In Juniper, `///` doc comments transform to GraphQL schema documentation. Haven't tried that in async-graphql, actually -- that'll be interesting to see.

leebenson replied:
Nice, `///` -> schema docs:

<img width="632" alt="Screenshot 2020-08-24 at 13 56 18" src="https://user-images.githubusercontent.com/719722/91047331-c6a8b880-e611-11ea-808a-aa21b37f2f2c.png">

JeanMertz commented on src/api/server.rs:
```diff
@@ -0,0 +1,123 @@
+    address: SocketAddr,
+
+    /// Transmission channel to trigger closure of a running API server
+    trigger_cancel: Sender<()>,
```
minor nit/suggestion/question:

I've been getting into the habit of referring to more "exotic" types through their parent module name instead of importing the type itself at the top of the file.

For example, in this case, you'd do `use tokio::sync::oneshot` at the top, and then `oneshot::Sender`. To me, this provides more context at the callsite _what kind of sender_ this is, at the cost of more verbose paths.

Same goes for other types, such as `http::Response` vs `Response`, etc.

There isn't really a hard-and-fast rule we can apply here, but wondering what others think. It's a nit though, so feel free to ignore this comment if you want.

leebenson replied:
👍 I like that, personally. I started to do that with `api::*`, rather than the end structs (although admittedly, not universally.) I do think it helps with readability, and also lessens the import noise.

I'll do a sweep to enforce that more uniformly.

JeanMertz commented on src/api/server.rs:
```diff
@@ -0,0 +1,123 @@
+use async_graphql_warp::{graphql_subscription, GQLResponse};
+use std::convert::Infallible;
+use std::net::SocketAddr;
+use tokio::sync::oneshot;
+use tokio::sync::oneshot::{Receiver, Sender};
```
```suggestion
use tokio::sync::oneshot::{self, Receiver, Sender};
```

JeanMertz commented on src/api/server.rs:
```diff
@@ -0,0 +1,123 @@
+
+impl Server {
+    /// Returns a new API Server
+    pub fn new(address: SocketAddr) -> Server {
```
```suggestion
    pub fn new(address: SocketAddr) -> Self {
```

(not sure how strictly Clippy is configured in Vector, but might be good to run Clippy locally as well, as it should catch cases like these)

JeanMertz commented on src/api/server.rs:
```diff
@@ -0,0 +1,123 @@
+    }
+
+    /// String representation of the bound port
+    pub fn port(&self) -> String {
```
Might make sense to just return `u16` here, and have the callee convert the port to a string if need be.

leebenson replied:
Thanks. I did this because the only current usage is logging, which is otherwise `port().to_string().as_str()`, which felt a bit aesthetically unpleasing. I think it's probably a better API to return the actual thing and let the callee transform it though, I think.

JeanMertz replied:
Sounds good to me 👍

> which is otherwise `port().to_string().as_str()`, which felt a bit aesthetically unpleasing

You should be able to dereference (`&`) the string, since `String` implements the `Deref` trait, which returns a `&str`. This makes it a bit more pleasant to read.

By the way, one more convention I learned a while back when digging into Rust is to use `to_owned()` instead of `to_string()` to go from a `&str` to a `String`.

Someone (somewhere) explained it as being more natural, as it feels a bit strange that you already have a string (`&str`) on which you call `to_string()`. `to_owned()` makes it very explicit that you are about to allocate memory to get an owned copy of the string.

This doesn't always work though, specifically in this case, you're going from `IpAddr` to `String`, which requires you to use the `ToString` trait, as `ToOwned` simply returns an owned `IpAddr`.

JeanMertz commented on src/api/server.rs:
```diff
@@ -0,0 +1,123 @@
+    }
+
+    /// String representation of the bound IP address
+    pub fn ip(&self) -> String {
```
It's more idiomatic to return a reference to a string (`&str`). This removes the forced allocation, but still allows the callee to decide if it needs an owned string by calling `to_string()` itself.

JeanMertz replied:
Ah, never mind. There is no way to get a string representation of `IpAddr` without allocating.

JeanMertz replied:
This could be a reason to have this function return the `IpAddr` type instead, and have the callee call `to_string` if needed. Similar to our discussion about `port()`.

But having said that, since this is only used in one location, I don't mind too much what choice we make.

leebenson replied:
Updated locally to return the underlying types, and require the callee to convert as required:

Usage is now:

```rust
info!(
    message = format!("{} API server", message).as_str(),
    ip = server.ip().to_string().as_str(),
    port = server.port().to_string().as_str()
);
```

The `.to_string().as_str()` hope is there to satisfy the `tracing::Value` trait bound, which isn't implement for `String`.

JeanMertz replied:
Looks good to me 👍

I suspect `port = &server.port().to_string()` should work as wel, as `String` derefences to `&str`.

See also: https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=741616df7f3cb47af85b219779d654fa

leebenson replied:
This is to satisfy `tracing::Value`:

```rust
error[E0277]: the trait bound `std::string::String: tracing::Value` is not satisfied
   --> src/topology/mod.rs:493:9
    |
493 | /         info!(
494 | |             message = format!("{} API server", message).as_str(),
495 | |             ip = server.ip().to_string(),
496 | |             port = server.port().to_string().as_str()
497 | |         );
    | |__________^ the trait `tracing::Value` is not implemented for `std::string::String`

JeanMertz replied:
> This is to satisfy `tracing::Value`

I understand, what I suggested was applying [deref coercion](https://doc.rust-lang.org/nomicon/coercions.html):

> Expression &x of type &T to &*x of type &U if T derefs to U (i.e. T: Deref<Target=U>)

since `impl Deref<Target=str> for String` [exists](https://doc.rust-lang.org/1.45.2/std/string/struct.String.html#impl-Deref), that gets rid of the extra `as_str()`. Anyway, it's a stylistic difference, so there's no need to change anything here, just wanted to point this out for future use 👍.

JeanMertz commented on src/api/server.rs:
```diff
@@ -0,0 +1,123 @@
+        Response::builder()
+            .header("content-type", "text/html")
+            .body(playground_source(
+                GraphQLPlaygroundConfig::new("/graphql").subscription_endpoint("/graphql"),
```
I was wondering if we should use a path that describes the data you can get at that endpoint, vs. the technology used to query that data? (f.e. `/metrics` vs `/graphql`?). But I'm not even sure that's necessarily better, and it's a minor nit regardless.

leebenson replied:
The current intention is to use GraphQL as the fuller interface for the API (and `vector top`), which is currently:

- Health / heartbeats
- Metrics (in progress)
- Reading / validating `Cargo.toml` (in progress)

The roadmap isn't super clear atm, so I'm not sure how much this will evolve, but it feels likely that metrics alone probably won't encompass the full UI <-> Vector comms.

What _is_ likely though, I think, is having multiple interfaces into the same API, which as regular HTTP endpoints for the same thing. Therefore, I think `/graphql` + `/<other_interfaces>` is a good place to start.

JeanMertz replied:
Yeah, that makes sense 👍

JeanMertz commented on src/config/diff.rs:
```diff
@@ -15,6 +20,8 @@ impl ConfigDiff {
     pub fn new(old: &Config, new: &Config) -> Self {
         ConfigDiff {
+            #[cfg(feature = "api")]
+            api: api::Diff::from_api(&old.api, &new.api),
```
nit: All the other ones are called `Difference`, so maybe use a similar name for `api::Diff`?

JeanMertz replied:
I'm also unsure if we _want_ to expose the api diff through the `ConfigDiff` type?

I don't know enough about this part of the code-base to reason one way or the other, but it does seem like right now it only exposes diffs for sources/transforms/sinks, while there are other top-level Vector configuration options as well that could differ, but aren't included in this type?

leebenson replied:
👍 Will rename to `Difference`.

I'm piggybacking `ConfigDiff` because it seemed like a catch-all for any configuration changes that affect topology.

I could introduce a separate type here, although I'm not clear if that would add much benefit; I like the fact we can detect config differences in one place rather than splitting it up.

What might be better though is not having to do this 'raw' match on `api::Diff` in topology, and instead have convenience methods on the diff itself to detect whether it's a Start|Restart|Stop. That would avoid needing to bring in separate types to discern changes.

leebenson replied:
Updated in 11b98056edd574979318b912f4a96bcaa9019f45.

Added some convenience methods to save some matching on the 'inner' `api::Difference` enum, in places:

```rust
#[cfg(feature = "api")]
let api = match &diff.api {
    Some(api_diff) if api_diff.is_start_or_restart() => {
        Some(Server::new(config.api.bind.unwrap()))
    }
    _ => None,
};
```

JeanMertz reviewed (commented):
Sweet @leebenson 🎉

I left some comments/nits, but overall this looks like a great start.

leebenson commented on src/api/server.rs:
```diff
@@ -0,0 +1,123 @@
+    );
+
+    // GraphQL playground. Also sets up /graphql as an endpoint for both queries + subscriptions
+    let graphql_playground = warp::path("playground").map(|| {
```
I was thinking it might be nice to have this configurable, maybe...

```
[api]
  playground = true
```

Thinking it might be useful to have something to query directly, outside of the main UI.

What do you think?

JeanMertz replied:
It seems like it could bloat the binary, but I'm not sure how much data is embedded to support the playground (although this implies we'd need a way to remove that data from being embedded by default, which might not be possible with the used GraphQL library).

Maybe the assets are fetched from a remote server? If so, that would introduce another potential reason not to embed it: we can't be sure systems have internet access, so this feature might not work in those cases.

I'm happy to defer this though, it's easy to change later, while we work out the exact details, I can at least see how this is useful during debugging 👍

leebenson replied:
Had a quick look at this -- the embedded `index.html` file is 21,840 bytes.

Static assets (JS, etc) are pulled remotely on usage:

```html
  <link rel="stylesheet" href="//cdn.jsdelivr.net/npm/graphql-playground-react/build/static/css/index.css" />
  <link rel="shortcut icon" href="//cdn.jsdelivr.net/npm/graphql-playground-react/build/favicon.png" />
  <script src="//cdn.jsdelivr.net/npm/graphql-playground-react/build/static/js/middleware.js"></script>
```

There doesn't seem to be a way to disable embedding it -- [it's returned as a string here](https://github.com/async-graphql/async-graphql/blob/7efeeb543b4a9ee02f03e13cf394bf5e40d6b865/src/http/playground_source.rs) when calling `playground_source`.

This would have been better as a macro IMO, but at 21k I'm not sure it warrants too much additional effort.

I think the more import factor is allowing the user to turn it on/off. More from the standpoint of letting users decide whether they want to host a HTML server on the API port.

leebenson commented:
Thanks for taking a look at this, @JeanMertz. Appreciated!

fanatid committed:
chore(deps): update bollard from 0.7.1 to 0.8.0 (#3534)

Signed-off-by: Kirill Fomichev <fanatid@ya.ru>

fanatid committed:
chore(deps): bump rusoto from 0.44.0 to 0.45.0 (#3527)

Signed-off-by: Kirill Fomichev <fanatid@ya.ru>

StephenWakely committed:
fix(statsd sink): Handle Absolute kind for values other than Gauge. (#3510)

* StatsD sink should handle Absolute Values for values other than Gauge.

Signed-off-by: Stephen Wakely <fungus.humungus@gmail.com>

binarylogic committed:
docs: Correct in_flight_limit default in docs

Signed-off-by: binarylogic <bjohnson@binarylogic.com>

jamtur01 committed:
docs(external docs): Some grammar and structure edits to .meta content (#3548)

Signed-off-by: James Turnbull <james@lovedthanlost.net>

binarylogic committed:
chore: Add .meta/links.toml

Signed-off-by: binarylogic <bjohnson@binarylogic.com>

binarylogic committed:
chore: Regenerate vector.spec.toml

Signed-off-by: binarylogic <bjohnson@binarylogic.com>

dependabot[bot] committed:
chore(deps): bump getset from 0.1.0 to 0.1.1 (#3296)

Bumps [getset](https://github.com/Hoverbear/getset) from 0.1.0 to 0.1.1.
- [Release notes](https://github.com/Hoverbear/getset/releases)
- [Commits](https://github.com/Hoverbear/getset/commits)

Signed-off-by: dependabot[bot] <support@github.com>

Co-authored-by: dependabot[bot] <49699333+dependabot[bot]@users.noreply.github.com>

jamtur01 committed:
Added support for the ci-condition: skip label (#3539)

As GitHub Actions doesn't support skipping whole workflows and you can't set
job defaults we need to specify the skip on every job.

Signed-off-by: James Turnbull <james@lovedthanlost.net>

jamtur01 committed:
Update semantic.yml

Added `kubernetes_logs source`.

markonen committed:
feat(geoip transform): ASN/ISP database support (#3265)

Co-authored-by: binarylogic <bjohnson@binarylogic.com>

ktff committed:
enhancement(file sink): Sync all data before finishing (#3475)

* Add sync_all

Signed-off-by: ktf <krunotf@gmail.com>

* Add close

Signed-off-by: Kruno Tomola Fabro <krunotf@gmail.com>

leebenson committed:
graphql metrics

leebenson committed:
graphql metrics

fanatid committed:
chore: add more tokio::test (#3505)

Signed-off-by: Kirill Fomichev <fanatid@ya.ru>

bruceg committed:
enhancement(json_parser transform): Enhance instrumentation (#3521)

* enhancement(json_parser transform): Enhance instrumentation

Signed-off-by: Bruce Guenter <bruce@timber.io>

bruceg committed:
enhancement(socket source): Add and unify events (#3486)

* Fix up one socket source event

* Add event for UDP socket send error

* Consolidate most of the TCP/UDP/Unix events

Signed-off-by: Bruce Guenter <bruce@timber.io>

leebenson committed:
Merge branch 'master' into leebenson/api

leebenson committed:
Merge branch 'master' into leebenson/api

leebenson committed:
remove context

leebenson committed:
remove context

MOZGIII committed:
chore: Add k8s related code parts to CODEOWNERS (#3562)

Signed-off-by: MOZGIII <mike-n@narod.ru>

MOZGIII committed:
chore(file source): Multiline config extract (#3466)

* Extract the the multiline config to allow sharing among the sources

Signed-off-by: MOZGIII <mike-n@narod.ru>

* Add multiline_config to CODEOWNERS

Signed-off-by: MOZGIII <mike-n@narod.ru>

bruceg committed:
enhancement(regex_parser transform): Enhance instrumentation (#3523)

* enhancement(regex_parser transform): Enhance instrumentation

* Move emit of RegexParserFailedMatch to proper location

If a conversion failed, `CompiledRegex::captures` would emit
`RegexParserConversionFailed` and return `None`. Its caller would then
also erroneously emit `RegexParserFailedMatch`, reporting a double
error. This is resolved by doing both emits in `fn captures`.

Signed-off-by: Bruce Guenter <bruce@timber.io>

jamtur01 committed:
fix: Fixed language on code fence in CONTRIBUTING.md (#3560)

leebenson committed:
metrics interface

leebenson committed:
metrics interface

Hoverbear committed:
chore(add_fields transform):  Add events to add_fields transform  (#3503)

* chore(add_fields transform): Add metrics for add_fields

Signed-off-by: Ana Hobden <operator@hoverbear.org>

* Use refs

Signed-off-by: Ana Hobden <operator@hoverbear.org>

* fmt

Signed-off-by: Ana Hobden <operator@hoverbear.org>

bruceg committed:
enhancement(grok_parser transform): Add internal events (#3553)

Signed-off-by: Bruce Guenter <bruce@timber.io>

jamtur01 committed:
fix: Remove quiet verbose options as environment variables to stop test-shutdown failing (#3559)

leebenson committed:
Merge branch 'master' into leebenson/api

leebenson committed:
Merge branch 'master' into leebenson/api

leebenson committed:
tidy up

leebenson committed:
tidy up

leebenson committed:
Update src/api/server.rs

Co-authored-by: Jean Mertz <git@jeanmertz.com>

leebenson committed:
Update src/api/server.rs

Co-authored-by: Jean Mertz <git@jeanmertz.com>

leebenson committed:
conditional playground

leebenson committed:
conditional playground

leebenson committed:
Update server.rs

leebenson committed:
Update server.rs

leebenson commented:
The GraphQL playground can now be disabled with:

```toml
[api]
  enabled = true
  playground = false
```

I've also added [the macro here](https://github.com/seanmonstar/warp/issues/619), to balance Warp `.or()` composition, which should improve build times.

In debug, this is wrapped in `warp::Filter::boxed` for dynamic dispatch; in release, this provides a more balanced or tree for static dispatch.

leebenson committed:
tidy up schema

leebenson committed:
tidy up schema

bruceg committed:
fix(performance): Only update auto-concurrency RTT from OK responses (#3493)

* fix(performance): Only update auto-concurrency RTT from ok responses

Error responses from a service may happen much faster (or rarely much
slower) than a successful response. This can dramatically skew the
measurements, leading to incorrect changes to the concurrency limit.

* Increase range of RTT test measurements

To allow for more internal skew on systems like Windows

* Return Option<f64> from EWMA average, to avoid use of uninitialized values

* Move concurrency limit management logic into its own function

Signed-off-by: Bruce Guenter <bruce@timber.io>

fanatid committed:
chore: more async functions in tests (#3578)

Signed-off-by: Kirill Fomichev <fanatid@ya.ru>

leebenson committed:
bump async-graphql; stitch schema

leebenson committed:
bump async-graphql; stitch schema

leebenson committed:
Merge branch 'master' into leebenson/api

leebenson committed:
Merge branch 'master' into leebenson/api

leebenson committed:
config test; fix lint

leebenson committed:
config test; fix lint

leebenson committed:
config test

leebenson committed:
config test

leebenson committed:
ok /health

leebenson committed:
ok /health

leebenson committed:
health test redux

leebenson committed:
health test redux

leebenson committed:
Merge branch 'leebenson/api' of github.com:timberio/vector into leebenson/api

leebenson committed:
Merge branch 'master' into leebenson/api

leebenson committed:
Create Cargo.lock

leebenson committed:
Merge branch 'master' into leebenson/api

leebenson committed:
mod tests

Signed-off-by: Lee Benson <lee@leebenson.com>

leebenson committed:
clippy

Signed-off-by: Lee Benson <lee@leebenson.com>

leebenson committed:
fix kafka

Signed-off-by: Lee Benson <lee@leebenson.com>

leebenson committed:
rustfmt

Signed-off-by: Lee Benson <lee@leebenson.com>

leebenson commented on src/test_util/mod.rs:
```diff
@@ -309,22 +309,32 @@ where
 }
 
-pub async fn wait_for<F, Fut>(mut f: F)
+pub async fn wait_for<F, Fut>(mut f: F, wait_for_secs: u64)
```
Updated this to take a `wait_for_secs`, to make this slightly more configurable as I was diagnosing a race condition. The default of 5 seconds turned out to be okay, but I left this in to give that extra control to the caller.

fanatid replied:
Maybe more flexible [std::time::Duration](https://doc.rust-lang.org/std/time/struct.Duration.html) instead secsonds?

leebenson replied:
Good idea!

leebenson commented on src/test_util/mod.rs:
```diff
@@ -309,22 +309,32 @@ where
 }
 
+// Wait for TCP socket to accept traffic-- defaults to 5 seconds
 pub async fn wait_for_tcp(addr: SocketAddr) {
```
To avoid changing the signature of any functions already calling `wait_for_tcp`, I defaulted this to 5 seconds to mimic the current behavior.

leebenson commented on src/test_util/mod.rs:
```diff
@@ -309,22 +309,32 @@ where
+}
+
+// Wait for TCP socket to accept traffic-- user definable seconds
+pub async fn wait_for_tcp_in_secs(addr: SocketAddr, secs: u64) {
```
Added this as a configurable version of `wait_for_tcp`.

leebenson commented on src/topology/mod.rs:
```diff
@@ -453,6 +471,41 @@ impl RunningTopology {
+
+        info!(
+            message = format!("{} API server", message).as_str(),
+            ip = server.ip().to_string().as_str(),
```
This seems a little verbose, but `tracing::Value` is required on a value, so the usual `String` -> `&str` deref didn't seem to work.

leebenson commented on src/topology/mod.rs:
```diff
@@ -453,6 +471,41 @@ impl RunningTopology {
+            port = server.port().to_string().as_str()
+        );
+
+        self.api = Some(server.run().await);
```
This takes ownership of the running `Server`, which has a `.stop()` method for terminating. I thought it might be more useful to do it this way vs. returning a function or shutdown signal that terminates, as then it allows for the possibility of calling `Server.ip()` or similar methods, for logging.

leebenson commented on tests/api.rs:
```diff
@@ -0,0 +1,37 @@
+        let (mut _top, _) = start_topology(config, false).await;
+        wait_for_tcp_in_secs(addr, 30).await;
+
+        let url = format!("http://{}:{}/health", addr.ip(), addr.port());
```
GET check on `/health`

leebenson commented on src/api/schema/metrics.rs:
```diff
@@ -0,0 +1,88 @@
+
+/// Returns a stream of `Metric`s, collected at the provided millisecond interval
+fn get_metrics(interval: i32) -> impl Stream<Item = Metric> {
+    let controller = get_controller().unwrap();
```
Rather than treat the API as a sink, I thought it made sense to utilise the 'global' controller directly. This is called per-subscription, at the point of query, and returns a stream that's specific to that one subscription.

I'm using the `stream!` macro to yield the inner value when the value is of type `Event::Metric`.

lukesteensen replied:
The main reason I'd caution against this is that it ties us pretty directly to the current implementation of metrics and doesn't give us a clear path to exposing non-metric data. The idea of making internal metrics a source is that they become less special (e.g. represented by the same data type as user metrics) and any improvements we make to vector's internal observability tools should help improve our users' ability to handle their own metrics.

Basically, internal metrics as a source was an intentional design decision so let's not sidestep it.

leebenson commented on src/api/schema/metrics.rs:
```diff
@@ -0,0 +1,88 @@
+        #[arg(validator(IntRange(min = "100", max = "60_000")))] interval: i32,
+    ) -> impl Stream<Item = MetricType> {
+        get_metrics(interval).filter_map(|m| match m.name.as_str() {
+            "events_processed" => Some(MetricType::EventsProcessed(m.into())),
```
Currently, only `events_processed` metrics are consumed. This will evolve to accommodate all metrics types, and will likely be consumed via specific queries rather than the generalised `get_metrics` subscription. I also imagine we'll be able to do additional filtering by params.

Since the design of that is TBD and likely to be covered by RFCs, this is just a meantime demonstration that it works.

leebenson reviewed (commented):
I think this is ready for review. The git history is a little screwy, due to an attempt to GPG sign commits without forcing the branch, which resulted in dupes. This will be squashed so it should be harmless.

The main features in this PR are:

- The initial plumbing for an API server, which (by default) runs on port 8686 when enabled.
- Configuration settable via the `[api]` block, in `vector.toml`
- `/health` endpoint
- `/graphql` query/subscription handler
- `/playground`, on by default (can be turned off with `api.playground = false`.

Along with the query:

```gql
{ health }
```

And subscriptions for a heartbeat, and metrics via:

```gql
subscription {
  metrics(interval: 1000) {
    timestamp
    ...on EventsProcessed {
      eventsProcessed
    }
  }
}
```

The initial GraphQL schema is:

```gql
directive @ifdef on FIELD
type BytesProcessed implements MetricType {
  timestamp: DateTimeUtc
  bytesProcessed: Float!
}

scalar DateTimeUtc

type EventsProcessed implements MetricType {
  timestamp: DateTimeUtc
  eventsProcessed: Float!
}

type Heartbeat {
  utc: DateTimeUtc!
}

interface MetricType {
  timestamp: DateTimeUtc
}

type Query {
  health: Boolean!
}

type Subscription {
  metrics(interval: Int!): MetricType!
  heartbeat(interval: Int! = 1000): Heartbeat!
}
```

This will evolve quickly, and shouldn't be taken too literally at this stage. I would avoid surfacing this in API docs yet as it needs a period of development / settling-in internally.

Most tests appear to be passing. One that isn't is the .deb package, but that appears to be due to `leveldb-sys`. Anything else that comes up related to this PR, I'll try to squash in the interim.

I've added two integration tests:

1. Bolting on to `tests/config.rs`, to test for the `[api]` configuration
2. A new `tests/api.rs`, which currently just spawns topology and pings the `/health` endpoint.

I left out any GraphQL testing in this PR, as it adds some complexity testing WebSockets that I felt was a distraction given the evolving nature of the schema. Once the schema settles, I intend on adding a full suite of tests for every query/subscription.

As noted in the PR, there are a stack of improvements that can be added later that I foresee evolving into-- TLS config for the API, user-definable endpoints, that kinda thing. This is just a starting point.

This is my first Rust stint, and the first time I've had the opportunity to get acquainted with the Vector codebase, so very much open to any/all feedback/pointers. Thank you!

MOZGIII commented on src/api/server.rs:
```diff
@@ -0,0 +1,134 @@
+                        GraphQLPlaygroundConfig::new("/graphql").subscription_endpoint("/graphql"),
+                    ))
+            } else {
+                Response::builder()
```
Would this work?
```rust
let mut routes = balanced_or_tree!(...); // non-conditional routes
if enable_playground {
  routes = routes.or(graphql_playground);
}
routes.with(...) // the rest 0f the initialization
```

leebenson replied:
This is the error:

```rust
error[E0308]: mismatched types
   --> src/api/server.rs:121:22
    |
106 |               let graphql_playground = warp::path("playground").map(move || {
    |  ___________________________________________________________________-
107 | |                 if enable_playground {
108 | |                     Response::builder()
109 | |                         .header("content-type", "text/html")
...   |
118 | |                 }
119 | |             });
    | |_____________- the found closure
120 | 
121 |               routes = routes.or(graphql_playground);
    |                        ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ expected struct `warp::filter::boxed::BoxedFilter`, found struct `warp::filter::or::Or`
    | 
   ::: /Users/leebenson/.cargo/registry/src/github.com-1ecc6299db9ec823/async-graphql-warp-1.17.3/src/lib.rs:162:19
    |
162 |   ) -> BoxedFilter<(impl Reply,)>
    |                     ----------
    |                     |
    |                     one of the expected opaque types
    |                     one of the found opaque types
    | 
   ::: src/api/handler.rs:11:33
    |
11  |   pub async fn health() -> Result<impl Reply, Rejection> {
    |                                   ----------
    |                                   |
    |                                   one of the expected opaque types
    |                                   one of the found opaque types
    |
    = note: expected struct `warp::filter::or::Or<warp::filter::or::Or<warp::filter::boxed::BoxedFilter<_>, warp::filter::boxed::BoxedFilter<(impl warp::reply::Reply,)>>, warp::filter::or::Or<warp::filter::boxed::BoxedFilter<(async_graphql_warp::GQLResponse,)>, warp::filter::boxed::BoxedFilter<(std::result::Result<http::response::Response<std::string::String>, http::error::Error>,)>>>`
               found struct `warp::filter::or::Or<warp::filter::or::Or<warp::filter::or::Or<warp::filter::boxed::BoxedFilter<_>, warp::filter::boxed::BoxedFilter<(impl warp::reply::Reply,)>>, warp::filter::or::Or<warp::filter::boxed::BoxedFilter<(async_graphql_warp::GQLResponse,)>, warp::filter::boxed::BoxedFilter<(std::result::Result<http::response::Response<std::string::String>, http::error::Error>,)>>>, warp::filter::map::Map<warp::filters::path::Exact<warp::filters::path::internal::Opaque<&str>>, [closure@src/api/server.rs:106:67: 119:14 enable_playground:_]>>`
```

MOZGIII replied:
I see. The issue here is that the `.or` combinator returns a new value with a different type, that wraps the original. So, in case `if` executes the type of the `routes.or(graphql_playground)` would be different from what rust had initially eluded at `let route`, and thus not be assignable.

We can work around it by using `Box` with a trait object in it - this way the types will simply be homomorphic `Box<dyn Filter>` in all branches, thus resolving the type conflict.

Warp offers some neat helpers, so what we should be able to do is:

```rust
let mut routes = balanced_or_tree!(...).boxed(); // non-conditional routes
if enable_playground {
  routes = routes.or(graphql_playground).boxed();
}
routes.with(...) // the rest 0f the initialization
```

MOZGIII replied:
Apparently there's even no need to call `.boxed()` for `balanced_or_tree!(...)` (cause macro already does that internally) - just for `routes.or(graphql_playground).boxed()`.

leebenson replied:
I believe`.boxed()` is only called on debug builds, deferring to dynamic dispatch:

```rust
#[cfg(debug_assertions)]
macro_rules! debug_boxed {
    ($x:expr) => {
        ::warp::Filter::boxed($x)
    };
}

#[cfg(not(debug_assertions))]
macro_rules! debug_boxed {
    ($x:expr) => {
        $x
    };
}
```

Doesn't boxing it make routing slightly slower? I'm not sure if the overhead is enough to make a difference either way. I'd prefer to avoid the route if the performance penalty is negligible.

MOZGIII replied:
This is a good question, and I really don't know the answer. It seems like in our case the effect of it to performance won't matter, but, as always, this needs micro benches to say for sure.
There seems to be an alternative to dynamic dispatch in wrap though - the `unify` call. https://docs.rs/warp/0.2.4/warp/trait.Filter.html#method.unify

I'm not sure, but it looks like it could be used like this:

```rust
let routes = balanced_or_tree!(...); // non-conditional routes
let routes = if enable_playground {
  routes.or(graphql_playground).unify()
} else {
  routes.unify()
};
```

leebenson replied:
👍 Thanks, I'll take a look at that!

MOZGIII replied:
I updated the post above - just figured it's probably not for our use case. Also, see the commend re `boxed` :smile:

leebenson replied:
Haven't trouble getting the boxed types to work:

```rust
error[E0308]: mismatched types
   --> src/api/server.rs:63:18
    |
63  |         routes = routes.or(playground).boxed();
    |                  ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ expected opaque type, found enum `warp::generic::Either`
```

If you have the free cycles, would you mind updating this branch with a demo please? I'm getting lost in nested types.

MOZGIII replied:
Sure!

MOZGIII replied:
Take a look at https://github.com/timberio/vector/pull/3618

leebenson replied:
Thanks @MOZGIII, will take a look now

MOZGIII commented on src/api/server.rs:
```diff
@@ -0,0 +1,134 @@
+    }
+
+    /// Run the API server
+    pub async fn run(mut self) -> Self {
```
It's usually good to separate the "initialization routine" from the "runtime routine".

I.e. have a way to do everything that's going on this fn and wait for it to complete - and have another fn to start the `tokio::spawn(server);`.

It would make sense to have separate structs as initialization routine arguments (i.e. `Config`) to avoid for the function to take `self` and return `self` - but the second `self` missing the `cancel_channel` and not suitable for calling `run` on it anymore.

leebenson replied:
Thanks, makes sense. I'll work on improving this. I'd also like to get the global shutdown coordinator to signal closure, rather than rely on a separate channel. I'll play about with it some more and show you know what I come up with.

MOZGIII replied:
You can take inspiration from how file server does it: https://github.com/timberio/vector/blob/1e180bff8f1635b7d24cd9f4838af63255ec5137/lib/file-source/src/file_server.rs#L68
It takes an abstract future from outside as a shutdown signal and terminates when the future resolves.

MOZGIII commented on src/api/server.rs:
```diff
@@ -0,0 +1,134 @@
+    }
+
+    pub fn set_playground(mut self, enable: bool) -> Self {
+        self.playground = enable;
```
Consider using https://docs.rs/getset/0.1.1/getset/index.html and https://docs.rs/derive_builder/0.9.0/derive_builder/

JeanMertz commented on src/config/api.rs:
```diff
@@ -0,0 +1,82 @@
+}
+
+pub fn default_bind() -> Option<SocketAddr> {
+    Some(SocketAddr::new(Ipv4Addr::new(127, 0, 0, 1).into(), 8686))
```
I wondered if this port showed up on the ["well-known ports" list](https://en.wikipedia.org/wiki/List_of_TCP_and_UDP_port_numbers), turns out it doesn't 🎉

(we might be able to add it at some point 🥇)

edit: it does show up in [this](https://www.iana.org/assignments/service-names-port-numbers/service-names-port-numbers.txt) IANA list, but I wouldn't worry about that:

> sun-as-jmxrmi      8686        tcp    Sun App Server - JMX/RMI                                           2005-11
sun-as-jmxrmi      8686        udp    Sun App Server - JMX/RMI                                           2005-11

leebenson replied:
https://timberio.slack.com/archives/CCG0F0NJE/p1597672642095900

MOZGIII commented on tests/api.rs:
```diff
@@ -0,0 +1,37 @@
+        let addr = config.api.bind.unwrap();
+
+        let (mut _top, _) = start_topology(config, false).await;
+        wait_for_tcp_in_secs(addr, 30).await;
```
It is great to have the test, though I'm not a big fan of timers.
It would be cool if the topology exposed a channel for when the API setup is complete and the socket is listening - we could eliminate the race here, and potentially also utilize this for health checks.

leebenson replied:
Yeah, I just reached for this since it was already in use.

MOZGIII replied:
I think we're generally moving away from using timers in tests where possible. They caused us some unncessary maintenance effort in the past.

leebenson replied:
Happy to switch this to something else, but didn't want to touch topology too much in this first PR, in case it impacted other areas I'm not very familiar with.

Hoverbear commented on Cargo.toml:
```diff
@@ -131,7 +135,7 @@ uuid = { version = "0.7", features = ["serde", "v4"], optional = true }
 snafu = { version = "0.6", features = ["futures-01", "futures"] }
 url = "2.1.1"
-base64 = { version = "0.10.1", optional = true }
+base64 = { version = "0.12.3", optional = true }
```
Please always feel welcome to upgrade us to stable, published versions of existing dependencies. :)

Hoverbear commented on src/api/macros.rs:
```diff
@@ -0,0 +1,68 @@
+// - balanced_or_tree!(@internal a, b ; c, d, e ; e) // now only one elem in counter
+// - balanced_or_tree!(a, b, c).or(balanced_or_tree(d, e)) // recurse on each sublist
+#[macro_export]
+macro_rules! balanced_or_tree {
```
What kind of improvement can we expect from introducing this macro?

leebenson replied:
This is [the issue](https://github.com/seanmonstar/warp/issues/619) that motivated it. Quote:

>  I was able to cut debug build times by ~65%, from 6 minutes to 2 minutes, by building a more balanced Or tree of filters

Thought it was worth a crack.

Hoverbear replied:
Did you test this on vector?

Hoverbear commented on src/topology/mod.rs:
```diff
@@ -453,6 +471,41 @@ impl RunningTopology {
+        let server = new_pieces.api.take().unwrap();
+
+        info!(
+            message = format!("{} API server", message).as_str(),
```
It'd be really nice if we could include the URL as a copy/pastable string.

Hoverbear commented:
I presume subscriptions are for a future PR?

Otherwise, looking good. :)

![image](https://user-images.githubusercontent.com/130903/91593654-69fd1480-e915-11ea-9e45-96c646f89101.png)

![image](https://user-images.githubusercontent.com/130903/91593737-87ca7980-e915-11ea-9621-9837ad5dccba.png)

Hoverbear reviewed (changes requested):
Generally it would be better to add the `generator` events in a different PR, but given the simplicity of that source I don't see an issue.

This code is lacking documentation adequate to satisfy the review requirements listed: https://github.com/timberio/vector/blob/master/REVIEWING.md#documentation.

Otherwise, I think this is a very good first implementation step of this great epic!

leebenson commented:
@Hoverbear, thanks for the review!

Subscriptions should work. I can't tell from your screenshots, but it may be that your client isn't attempting a WebSocket handshake and is instead defaulting to a regular POST/GET? A query with a `subscription` block should do a WS upgrade.

Hoverbear commented:
So I should do like this?

```graphql
query foo {
  subscription bar {
    ...
  }
}
```

leebenson commented:
The original `subscription` block you have is correct:

```gql
subscription {
  heartbeat(interval: 1000) {
    utc 
  }
}
```

The issue is likely that whichever GraphQL client you're using isn't switching to WebSockets when it detects a `subscription` block.

To give you an example of how we'd typically distinguish between regular queries and subscriptions in the UI, we'd split logic based on the query type, like this:

```ts
return new ApolloClient({
  cache,
  link: ApolloLink.from([
    // Split on HTTP and WebSockets
    split(  // <-- this is a helper from the Apollo Client to branch logic based on a boolean test
      ({ query }) => {
        const definition = getMainDefinition(query);
        return (
          definition.kind === "OperationDefinition" &&
          definition.operation === "subscription" // <-- determines the operation type is 'subscription'
        );
      },
      // Use WebSockets for subscriptions
      new WebSocketLink(ws) as any,
      // ... fall-back to HTTP for everything else
      authHttp as any
    ),
  ]),
  queryDeduplication: false,
});
```

WebSockets are the de facto subscription protocol, but since it's opt-in and not always implemented, it often takes some configuration in the client.

The [GraphQL Playground](https://www.electronjs.org/apps/graphql-playground) electron app is one that should work out-the-box. Alternatively, you can go to `https://localhost:8686/playground`, which this PR includes. The electron app is a slightly more recent version which has 'tail' scrolling and is slightly easier to test with subscriptions, but otherwise it's identical.

When the schema stabilises, we'll wind up putting this into the [vector-ui](https://github.com/timberio/vector-ui) playground too, so we'll be able to test this stuff out without having to mess around with third-party clients and within our own UI.

Hoverbear commented:
@leebenson Perfect, thanks. Great answer!

MOZGIII committed:
Check if playground is enabled statically (#3618)

Signed-off-by: MOZGIII <mike-n@narod.ru>

leebenson committed:
fix 404

Signed-off-by: Lee Benson <lee@leebenson.com>

leebenson committed:
tidy routes

Signed-off-by: Lee Benson <lee@leebenson.com>

leebenson committed:
tidy not_found

Signed-off-by: Lee Benson <lee@leebenson.com>

leebenson committed:
fix route comment

Signed-off-by: Lee Benson <lee@leebenson.com>

lukesteensen commented on src/api/server.rs:
```diff
@@ -0,0 +1,134 @@
+    /// Address for the API server to bind on
+    address: SocketAddr,
+
+    /// Transmission channel to trigger closure of a running API server
+    trigger_cancel: oneshot::Sender<()>,
+
+    /// Receiver signal to cancel running API server
+    cancel_signal: Option<oneshot::Receiver<()>>,
```
We should be able to use our `ShutdownSignal` implementation for these.

lukesteensen commented on src/config/diff.rs:
```diff
@@ -2,7 +2,12 @@ use super::Config;
+
 pub struct ConfigDiff {
+    #[cfg(feature = "api")]
+    pub api: Option<api::Difference>,
```
I would honestly drop all of this `Difference` stuff for now. We can stick an on/off switch on `GlobalOptions` and leave it at that for now. I don't care about people being able to turn this on and off dynamically.

lukesteensen commented on src/config/mod.rs:
```diff
@@ -24,11 +24,17 @@ pub use diff::ConfigDiff;
     pub global: GlobalOptions,
+    #[cfg(feature = "api")]
+    #[serde(default)]
+    pub api: api::Options,
```
Again, I'd prefer we push this down into `GlobalOptions`.

bruceg commented on src/api/schema/metrics.rs:
```diff
@@ -0,0 +1,88 @@
+    let mut interval = tokio::time::interval(Duration::from_millis(interval as u64));
+
+    stream! {
+        while let _ = interval.tick().await {
```
This appears to be an infinite loop (`Interval::tick` cannot fail), so:
```suggestion
        loop {
            interval.tick().await;
```

lukesteensen commented on src/topology/mod.rs:
```diff
@@ -31,6 +34,8 @@ type TaskHandle = tokio::task::JoinHandle<Result<(), ()>>;
 #[allow(dead_code)]
 pub struct RunningTopology {
+    #[cfg(feature = "api")]
+    api: Option<Server>,
```
I would really prefer pretty much all of the changes in this file not happen :smile: (other than the spelling fixes, which should get their own PR).

We're currently working to try to simplify the topology code as much as possible and I just don't see any real benefit to putting it in change of the API server. We should start with something as simple as possible and just spawn the server in `main` when it's configured.

leebenson replied:
I currently lack a lot of context about what's stable or subject to change, and still fumbling my way through the code base, so bear with my lack of understanding here... but if the topology isn't updated, I'm not clear how we'd respond to config changes.

If someone updates:

```toml
[api]
  bind = "<new_ip_port>
```

... we'd want to respond to that the same as any other config, right?

lukesteensen commented on src/sinks/kafka.rs:
```diff
@@ -482,12 +482,15 @@ mod integration_test {
+                future::ready(high > 0)
+            },
+            5,
+        )
```
I'd prefer that we try to avoid changing unrelated code for the time being, just to help focus the PR.

lukesteensen reviewed (commented):
This is looking really interesting! For now, I'd really like to focus on the pros and cons of GraphQL, understanding its implementation and runtime characteristics, etc. It's hard to look at this and not think that we could do something simpler, so I'm curious how you weigh the complexity here.

bruceg commented on src/config/api.rs:
```diff
@@ -0,0 +1,82 @@
+use serde::{Deserialize, Serialize};
+use std::net::{Ipv4Addr, SocketAddr};
+
+#[derive(Default, Debug, Deserialize, Serialize, PartialEq)]
+pub struct Options {
+    #[serde(default = "default_enabled")]
+    pub enabled: bool,
+
+    #[serde(default = "default_bind")]
+    pub bind: Option<SocketAddr>,
+
+    #[serde(default = "default_playground")]
+    pub playground: bool,
+}
```
Would it make sense to use `#[derivative(Default)]` to have the same defaults for serde and within Rust?

bruceg commented on src/internal_events/generator.rs:
```diff
@@ -0,0 +1,14 @@
+#[derive(Debug)]
+pub struct GeneratorEventProcessed;
+
+impl InternalEvent for GeneratorEventProcessed {
```
I'll be rebasing that after some other things are merged, so no worries at this point.

bruceg commented on src/internal_events/mod.rs:
```diff
@@ -61,6 +62,7 @@ pub use self::blackhole::*;
 pub use self::docker::*;
 pub use self::elasticsearch::*;
 pub use self::file::*;
+pub use self::generator::*;
```
Should guard these with `#[cfg(feature = "sources-generator")]` and then you can use `pub(crate)` too.

bruceg commented on src/sources/docker.rs:
```diff
@@ -34,7 +34,7 @@ use string_cache::DefaultAtom as Atom;
 use tracing::field;
 
-/// The begining of image names of vector docker images packaged by vector.
+/// The beginning of image names of vector docker images packaged by vector.
```
These spelling fixes are good, but perhaps split them into another PR?

bruceg commented on src/sources/generator.rs:
```diff
@@ -87,6 +88,8 @@ impl GeneratorConfig {
                 .lines
                 .iter()
                 .map(|line| {
+                    emit!(GeneratorEventProcessed);
```
This might not be the best place, but it should work fine.

bruceg reviewed (commented):
I haven't taken the time to follow some of the API discussions in depth, so this is mostly just a surface scan.

A lot of the comments you have started conversations with in the PR would be good in-code explanations of why you did things that way for those who will read it after the PR is resolved. Consider adding them to the code.

leebenson commented:
Thanks for your review, @lukesteensen. Will address the points you raised more specifically on Monday, but I wanted to touch base on this real quick:

> For now, I'd really like to focus on the pros and cons of GraphQL, understanding its implementation and runtime characteristics, etc. It's hard to look at this and not think that we could do something simpler, so I'm curious how you weigh the complexity here.

I'd summarise the pros as:

- A strongly typed API, front and back. Clients know what's available, and the shape of data at compile time.
- Trivial to consume in the UI. This is perhaps less obvious in the back-end, but the UI benefits from this enormously. There's a glut of complexity that we side-step by opting for a well-defined spec and relatively well-understood tooling that will lighten the burden of what will undoubtedly be a very complex front-end app. I'd defer to @sghall's thoughts here too.
- A year's worth of experience in Alloy working with GraphQL in the UI + back-end (albeit Go) in Alloy (my personal experience with GraphQL goes back to 2015.), and a compatible toolchain in vector-ui.
- Payloads that lend themselves well to highly structured data.
- Regular POST semantics for one-time queries. No magic over HTTP.
- Subscriptions over WebSockets; a well-tested web protocol.
- JSON payloads.
- Relatively lightweight to overlay existing types and libs.
- Leverages existing libs; Warp in our case, for example.
- Strong tooling: A heavily used React client, [GraphQL Code Generator](https://graphql-code-generator.com/) to generate Typescript types + declarative queries/subscriptions, Chrome extensions, a built-in GraphQL Playground, etc.
- Heavy use in production across a variety of APIs; not niche.

And the cons as:

- Generally geared toward graph data, although this isn't a hard requirement.
- Queries with unknown levels of nesting can't be queried statically; generally, a need for queries with flatter structures arise.
- Versioning is trickier, given the lack of distinct endpoints. Nothing stopping multiple endpoints representing versions, although typically this is done in the schema.
- Some annoying rough edges -- the lack of input unions, for example, typically worked around with more specific queries.
- Reliance on a relatively young library rather than a 'pure' client like a stdlib tcp server.
- A need to invest, longer-term, in understanding and contributing to that external library.

I think it's important a strong emphasis is played on both the client, as much as the server, when considering how the data we're exposing is being consumed.

You comment about opting for something simpler, I think, is weighted toward Vector; the UI has to solve connection handling, data fetching, retries, state management, caching, propogation across multiple components, single and real-time queries, error handling and other concerns that we have to define an approach for, irrespective of protocol, with make many of these 'simpler approaches' anything but. GraphQL simplifies many of those areas by providing an ecosystem of tooling, and a well-defined spec, to avoid the need for DIY implementations.

leebenson commented:
@lukesteensen, just on this point:

> I don't care about people being able to turn this on and off dynamically.

There's a [Slack thread](https://timberio.slack.com/archives/C0194T26H38/p1598458970006300?thread_ts=1598457695.004700&cid=C0194T26H38) with some extra comments on that point that might be worth weighing in on.

If we're ultimately surfacing internal configs and event lines, which the UI designs seem to suggest will be the case, insecure-by-default and 'not able to flick the off switch' are positions we should consider carefully.

leebenson commented:
@lukesteensen - Apologies, GH isn't letting me inline respond to all comments, hence the break-out:

> The main reason I'd caution against this is that it ties us pretty directly to the current implementation of metrics and doesn't give us a clear path to exposing non-metric data. The idea of making internal metrics a source is that they become less special (e.g. represented by the same data type as user metrics) and any improvements we make to vector's internal observability tools should help improve our users' ability to handle their own metrics.
>
> Basically, internal metrics as a source was an intentional design decision so let's not sidestep it.

To be clear, this means of consuming metrics data isn't intended to limit the consumption of non-metric data. It doesn't, for example, prevent the API from reading topology data, or exposing `vector.toml` validation tooling, or later introducing the 'tap' command, which I imagined would be unrelated to the internal_metrics source.

It's simply a way for a an individual client subscription (which may exist in its own green thread, detached from all other subscribing clients) to return a stream that's unique to that one connection.

If the intention is to both detach the API from the topology module _and_ also consume it as a source, I'm not clear how those two objectives coincide. Treating the API as a quasi-sink seems to required it to be a component of toplogy, unless I'm missing something. Reading from the controller directly, per-subscription, and being able to granularly define an `interval` outside of the context of a sink, seemed to fit the API paradigm better.

Again, apologies if I'm misunderstanding something fundamental here, I'm both trying to learn how the code works and catch up on design decisions that may impact the future landscape, which isn't a balance I have right yet!

sghall commented:
@lukesteensen  I think it would be better for the GraphQL debate (if there is going to be another one) to go back to the RFC process.  As @leebenson pointed out, this has many implications for client/server communication, tooling and the way the UI would be built.  This IMO is outside the scope of this PR.  Everyone working on this project should be included in the process and have a chance to comment 👍

leebenson commented:
Closing this and moving to a separate PR to avoid a painful rebase. The API is being pulled out of topology anyway so v2 should be slimmer.
